src/utils/models.py

```diff
@@ -348,3 +354,170 @@ def fetch(url, up_auth_file='', up_base_url='', ojs_auth_file=''):
+        except (KeyError, AttributeError) as error:
+            self.fail(error)
+            return None
+
+    def fail(self, error):
+        self.stopped = timezone.datetime.now()
+        self.status = self.RORImportStatus.FAILED
+        self.save()
+        logger.error(error)
+        RORImportError.objects.create(ror_import=self, messsage=error)
+
+    @property
+    def ongoing(self):
+        return self.status == self.RORImportStatus.ONGOING
+
+    def get_records(self):
+        """
+        Gets the manifest of available data and checks if it contains
+        anything new. If there is no new data, or if the previous import failed,
+        the import is marked as unnecessary.
+        """
+        records_url = 'https://zenodo.org/api/communities/ror-data/records?sort=newest'
+        try:
+            response = requests.get(records_url, timeout=settings.HTTP_TIMEOUT_SECONDS)
+            response.raise_for_status()
+            self.records = response.json()
+            self.save()
+            if not self.new_download_needed:
+                self.status = self.RORImportStatus.ONGOING
+                self.save()
+        except requests.RequestException as error:
+            self.fail(error)
+
+    def delete_previous_download(self):
+        if not self.previous_import:
+            logger.debug('No previous import to remove.')
+            return
+        try:
+            os.unlink(self.previous_import.zip_path)
+        except FileNotFoundError:
+            logger.debug('Previous import had no zip file.')
+
+    def download_data(self):
+        """
+        Downloads the current data dump from Zenodo.
+        Then removes previous files to save space.
+        """
+        try:
+            response = requests.get(
+                self.download_link,
+                timeout=settings.HTTP_TIMEOUT_SECONDS,
+                stream=True,
+            )
+            response.raise_for_status()
+            with open(self.zip_path, 'wb') as zip_ref:
+                for chunk in response.iter_content(chunk_size=128):
+                    zip_ref.write(chunk)
+            if os.path.exists(self.zip_path):
+                self.delete_previous_download()
+        except requests.RequestException as error:
+            self.fail(error)
+
+    def filter_new_records(self, ror_data, organizations):
+        """
+        Finds new records from the data dump based on
+        ROR's last_modified time stamp.
+        """
+        timestamps = {}
+        for ror, ts in organizations.values_list("ror", "ror_record_timestamp"):
+            timestamps[ror] = ts
+        filtered_data = []
+        for record in tqdm.tqdm(ror_data, desc="Finding new or updated ROR records"):
+            ror = record.get("id", {})
+            last_modified = record.get("admin", {}).get("last_modified", {})
+            timestamp = last_modified.get("date", "")
+            if ror and timestamp and timestamp > timestamps.get(ror, ''):
```

Comment:
Can we make sure the defaults for these `.get()` are also floats for explicitness? we can go with `timestamps.get(ror, float())` or `timestamps.get(ror, float('-inf'))` if the timestamp could somehow be 0.0

Reply:
This timestamp will always be a string, actually. It's just `YYYY-MM-DD` from the ROR JSON, no Python object being created. Is that alright?